File: mobile_order/lib/models/quick_order.dart
// import 'package:meta/meta.dart';

class SelectedOption {
  final String name; // 옵션 이름 (예: "샷 추가")
  final String choice; // 사용자가 선택한 옵션 값 (예: "추가")
  final int price; // 선택지에 따른 추가 가격
  final int? quantity; // 선택지에 따른 수량

  SelectedOption({
    required this.name,
    required this.choice,
    required this.price, // 가격을 int로 설정
    this.quantity, // 수량은 nullable
  });

  factory SelectedOption.fromJson(Map<String, dynamic> json) {
    return SelectedOption(
      name: json['name'],
      choice: json['choice'],
      price: json['price'] as int, // int로 변환
      quantity: json['quantity'] as int?, // nullable로 변환
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'name': name,
      'choice': choice,
      'price': price,
      'quantity': quantity,
    };
  }
}

class QuickOrderItem {
  final String id;
  final String name;
  final int price;
  final int quantity;
  final bool isComplimentary;
  final List<SelectedOption> selectedOptions; // 선택된 옵션 목록
  const QuickOrderItem({
    required this.id,
    required this.name,
    required this.price,
    required this.quantity,
    this.isComplimentary = false,
    this.selectedOptions = const [], // 기본값으로 빈 목록
  });

  factory QuickOrderItem.fromJson(Map<String, dynamic> json) {
    return QuickOrderItem(
      id: json['id'] as String,
      name: json['name'] as String,
      price: json['price'] as int,
      quantity: json['quantity'] as int,
      isComplimentary: json['isComplimentary'] as bool? ?? false,
      selectedOptions: (json['selectedOptions'] as List<dynamic>?)
              ?.map((option) =>
                  SelectedOption.fromJson(option as Map<String, dynamic>))
              .toList() ??
          [],
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'name': name,
      'price': price,
      'quantity': quantity,
      'isComplimentary': isComplimentary,
      'selectedOptions':
          selectedOptions.map((option) => option.toJson()).toList(),
    };
  }

  QuickOrderItem copyWith({
    String? id,
    String? name,
    int? price, // double에서 int로 변경
    int? quantity,
    bool? isComplimentary,
  }) {
    return QuickOrderItem(
      id: id ?? this.id,
      name: name ?? this.name,
      price: price ?? this.price,
      quantity: quantity ?? this.quantity,
      isComplimentary: isComplimentary ?? this.isComplimentary,
    );
  }
}

class QuickOrder {
  final String id;
  final String restaurantId;
  final String businessDayId; // 추가된 필드
  final int? orderNumber;
  final List<QuickOrderItem> items;
  int queuePosition; // final 제거
  String status; // final 제거
  final int totalAmount;
  final bool isComplimentaryOrder;
  final DateTime createdAt;
  final DateTime updatedAt;
  final String user; // 사용자 참조를 위한 ID
  final String? fcmToken; // FCM 토큰 필드 추가
  QuickOrder({
    required this.id,
    required this.restaurantId,
    required this.businessDayId, // 생성자에 추가
    this.orderNumber,
    required this.items,
    this.queuePosition = 0,
    this.status = 'pending',
    required this.totalAmount,
    this.isComplimentaryOrder = false,
    required this.createdAt,
    required this.updatedAt,
    required this.user,
    this.fcmToken, // 생성자에 추가
  });

  factory QuickOrder.fromJson(Map<String, dynamic> json) {
    return QuickOrder(
      id: json['_id']?.toString() ?? json['id']?.toString() ?? '',
      restaurantId: json['restaurantId']?.toString() ?? '',
      businessDayId: json['businessDayId']?.toString() ?? '',
      orderNumber: json['orderNumber'] != null
          ? int.tryParse(json['orderNumber'].toString())
          : null,
      items: (json['items'] as List<dynamic>?)
              ?.map((item) =>
                  QuickOrderItem.fromJson(item as Map<String, dynamic>))
              .toList() ??
          [],
      queuePosition: int.tryParse(json['queuePosition']?.toString() ?? '') ?? 0,
      status: json['status']?.toString() ?? 'pending',
      totalAmount: int.tryParse(json['totalAmount']?.toString() ?? '') ?? 0,
      isComplimentaryOrder: json['isComplimentaryOrder'] as bool? ?? false,
      createdAt: json['createdAt'] != null
          ? DateTime.parse(json['createdAt'])
          : DateTime.now(),
      updatedAt: json['updatedAt'] != null
          ? DateTime.parse(json['updatedAt'])
          : DateTime.now(),
      user: json['user']?.toString() ?? '',
      fcmToken: json['fcmToken']?.toString(), // JSON 파싱에 추가
    );
  }

  Map<String, dynamic> toJson() {
    return {
      '_id': id,
      'restaurantId': restaurantId,
      'businessDayId': businessDayId, // JSON 변환에 추가
      'orderNumber': orderNumber,
      'items': items.map((item) => item.toJson()).toList(),
      'queuePosition': queuePosition,
      'status': status,
      'totalAmount': totalAmount,
      'isComplimentaryOrder': isComplimentaryOrder,
      'createdAt': createdAt.toIso8601String(),
      'updatedAt': updatedAt.toIso8601String(),
      'user': user,
      'fcmToken': fcmToken, // JSON 변환에 추가
    };
  }

  QuickOrder copyWith({
    String? id,
    String? restaurantId,
    String? businessDayId, // copyWith에 추가
    int? orderNumber,
    List<QuickOrderItem>? items,
    int? queuePosition,
    String? status,
    int? totalAmount,
    bool? isComplimentaryOrder,
    DateTime? createdAt,
    DateTime? updatedAt,
    String? user,
    String? fcmToken, // copyWith에 추가
  }) {
    return QuickOrder(
      id: id ?? this.id,
      restaurantId: restaurantId ?? this.restaurantId,
      businessDayId: businessDayId ?? this.businessDayId, // copyWith 메서드에 추가
      orderNumber: orderNumber ?? this.orderNumber,
      items: items ?? this.items,
      queuePosition: queuePosition ?? this.queuePosition,
      status: status ?? this.status,
      totalAmount: totalAmount ?? this.totalAmount,
      isComplimentaryOrder: isComplimentaryOrder ?? this.isComplimentaryOrder,
      createdAt: createdAt ?? this.createdAt,
      updatedAt: updatedAt ?? this.updatedAt,
      user: user ?? this.user,
      fcmToken: fcmToken ?? this.fcmToken, // copyWith에 추가
    );
  }
}

File: server/models/QuickOrder.js
//file: \server\models\QuickOrder.js

const mongoose = require("mongoose");
// 옵션 스키마

const SelectedOptionSchema = new mongoose.Schema({
  name: { type: String, required: true }, // 옵션 이름 (예: "샷 추가")
  choice: { type: String, required: true }, // 사용자가 선택한 옵션 값 (예: "추가")
  price: { type: Number, default: 0 }, // 선택지에 따른 추가 가격 (없을 경우 0)
  quantity: { type: Number }, // 선택지에 따른 수량 (없을 수 있음)
});
const QuickOrderItemSchema = new mongoose.Schema({
  id: { type: String, required: true },
  name: { type: String, required: true },
  price: { type: Number, required: true },
  quantity: { type: Number, required: true },
  selectedOptions: {
    type: [SelectedOptionSchema], // 옵션 배열
    default: [], // 옵션이 없을 경우 기본값으로 빈 배열
  },
});

const QuickOrderSchema = new mongoose.Schema({
  restaurantId: { type: String, required: true },
  businessDayId: { type: mongoose.Schema.Types.ObjectId, ref: "BusinessDay", required: true },
  orderNumber: { type: Number },
  items: [QuickOrderItemSchema],
  queuePosition: { type: Number, default: 0 },
  status: {
    type: String,
    enum: ["pending", "preparing", "served", "completed"],
    default: "pending",
  },
  totalAmount: { type: Number, required: true },
  isComplimentaryOrder: { type: Boolean, default: false }, // 추가
  createdAt: { type: Date, default: Date.now },
  updatedAt: { type: Date, default: Date.now },
  user: { type: String, required: true }, // ObjectId에서 String으로 변경
  fcmToken: { type: String }, // FCM 토큰 필드 추가
});

QuickOrderSchema.pre("save", async function (next) {
  if (this.isNew) {
    const BusinessDay = mongoose.model("BusinessDay");
    let businessDay = await BusinessDay.findById(this.businessDayId);

    if (!businessDay) {
      throw new Error("Business day not found");
    }

    // 해당 비즈니스 데이의 마지막 주문 번호를 가져옵니다.
    const lastOrder = await this.constructor
      .findOne({ businessDayId: this.businessDayId })
      .sort("-orderNumber");

    if (lastOrder) {
      // 같은 비즈니스 데이 내에서는 주문 번호를 증가시킵니다.
      this.orderNumber = lastOrder.orderNumber + 1;
    } else {
      // 새로운 비즈니스 데이의 경우, 주문 번호를 10부터 시작합니다.
      this.orderNumber = 10;
    }

    // BusinessDay 모델에 현재 주문 번호를 업데이트합니다.
    businessDay.currentOrderNumber = this.orderNumber;
    await businessDay.save();
  }
  next();
});

const QuickOrder = mongoose.model("QuickOrder", QuickOrderSchema, "quickOrders");

module.exports = QuickOrder;
